sdk/core/azure-core/src/http/websockets/websockets_impl.cpp

```diff
@@ -0,0 +1,884 @@
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Closed && m_state != SocketState::Invalid)
+    {
+      throw std::runtime_error("AddHeader can only be called on closed sockets.");
+    }
+    m_headers.emplace(std::make_pair(header, headerValue));
+  }
+
+  void WebSocketImplementation::Close(
+      uint16_t closeStatus,
+      std::string const& closeReason,
+      Azure::Core::Context const& context)
+  {
+    std::unique_lock<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+
+    // If we're closing an already closed socket, we're done.
+    if (m_state == SocketState::Closed)
+    {
+      return;
+    }
+    if (m_state != SocketState::Open)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+    m_state = SocketState::Closing;
+#if SUPPORT_NATIVE_TRANSPORT
+    if (m_transport->HasBuiltInWebSocketSupport())
+    {
+      m_transport->NativeCloseSocket(closeStatus, closeReason.c_str(), context);
+    }
+    else
+#endif
+    {
+      // Send a going away message to the server.
+      std::vector<uint8_t> closePayload;
+      closePayload.push_back(closeStatus >> 8);
+      closePayload.push_back(closeStatus & 0xff);
+      closePayload.insert(closePayload.end(), closeReason.begin(), closeReason.end());
+      std::vector<uint8_t> closeFrame = EncodeFrame(SocketOpcode::Close, true, closePayload);
+      SendTransportBuffer(closeFrame, context);
+
+      // Unlock the state mutex before waiting for the close response to be received.
+      lock.unlock();
+      // To ensure that we process the responses in a "timely" fashion, limit the close
+      // reception to 20 seconds if we don't already have a timeout.
+      Azure::Core::Context closeContext = context;
+      auto cancelTimepoint = closeContext.GetDeadline();
+      if (cancelTimepoint == Azure::DateTime::max())
+      {
+        closeContext = closeContext.WithDeadline(std::chrono::system_clock::now() + 20s);
+      }
+      // Drain the incoming series of frames from the server.
+      // Note that there might be in-flight frames that were sent from the other end of the
+      // WebSocket that we don't care about any more (since we're closing the WebSocket). So
+      // drain those frames.
+      auto closeResponse = ReceiveTransportFrame(context);
+      while (closeResponse && closeResponse->Opcode != SocketOpcode::Close)
+      {
+        m_receiveStatistics.FramesDroppedByClose++;
+        Log::Write(
+            Logger::Level::Warning,
+            "Received unexpected frame during close. Opcode: "
+                + std::to_string(static_cast<uint8_t>(closeResponse->Opcode)));
+        closeResponse = ReceiveTransportFrame(closeContext);
+      }
+
+      // Re-acquire the state lock once we've received the close lock.
+      lock.lock();
```

Comment:
What's the point of this lock.lock() statement?  It goes out of scope on the next line and will release.  Shouldn't this be after the `}`?

Reply:
The lock.lock() ensures that the lock is held when we get to line 260 (the call to shutdown the ping thread and close the transport and update m_state).

lock is declared at function scope, it doesn't go out of scope until line 264.